crates/sui-bridge-indexer/src/indexer_builder.rs

```diff
@@ -0,0 +1,405 @@
+        &self,
+        task_name: String,
+        starting_checkpoint: u64,
+        target_checkpoint: u64,
+    ) -> Result<(JoinHandle<Result<(), Error>>, Receiver<Vec<T>>), Error>;
+
+    fn get_block_number(data: &T) -> u64;
+}
+
+pub struct SuiCheckpointDatasource {
+    remote_store_url: String,
+    concurrency: usize,
+    checkpoint_path: PathBuf,
+    metrics: DataIngestionMetrics,
+}
+impl SuiCheckpointDatasource {
+    pub fn new(
+        remote_store_url: String,
+        concurrency: usize,
+        checkpoint_path: PathBuf,
+        metrics: DataIngestionMetrics,
+    ) -> Self {
+        SuiCheckpointDatasource {
+            remote_store_url,
+            concurrency,
+            checkpoint_path,
+            metrics,
+        }
+    }
+}
+
+#[async_trait]
+impl<P, R> Datasource<CheckpointTxnData, P, R> for SuiCheckpointDatasource
+where
+    P: Persistent<R> + 'static,
+    R: Sync + Send + 'static,
+{
+    async fn create_data_channel(
+        &self,
+        task_name: String,
+        starting_checkpoint: u64,
+        target_checkpoint: u64,
+    ) -> Result<
+        (
+            JoinHandle<Result<(), Error>>,
+            Receiver<Vec<CheckpointTxnData>>,
+        ),
+        Error,
+    > {
+        let (exit_sender, exit_receiver) = oneshot::channel();
+        let progress_store = SimpleProgressStore {
+            current_checkpoint: starting_checkpoint,
+            exit_checkpoint: target_checkpoint,
+            exit_sender: Some(exit_sender),
+        };
+        let mut executor = IndexerExecutor::new(progress_store, 1, self.metrics.clone());
+        let (data_sender, data_receiver) = metered_channel::channel(
+            1000,
+            &mysten_metrics::get_metrics()
+                .unwrap()
+                .channel_inflight
+                .with_label_values(&[&task_name]),
+        );
+        let worker = IndexerWorker::new(data_sender);
+        let worker_pool = WorkerPool::new(worker, task_name, self.concurrency);
+        executor.register(worker_pool).await?;
+        let checkpoint_path = self.checkpoint_path.clone();
+        let remote_store_url = self.remote_store_url.clone();
+        let join_handle = spawn_monitored_task!(async {
+            executor
+                .run(
+                    checkpoint_path,
+                    Some(remote_store_url),
+                    vec![], // optional remote store access options
+                    ReaderOptions::default(),
+                    exit_receiver,
+                )
+                .await?;
+            Ok(())
+        });
+        Ok((join_handle, data_receiver))
+    }
+
+    fn get_block_number((_, block_number, _): &CheckpointTxnData) -> u64 {
+        *block_number
+    }
+}
+
+pub enum BackfillStrategy {
+    Simple,
+    Partitioned { task_size: u64 },
+    Disabled,
+}
+
+pub trait DataMapper<T, R>: Sync + Send {
+    fn map(&self, data: T) -> Result<Vec<R>, anyhow::Error>;
+}
+
+struct SimpleProgressStore {
```

Comment:
one suggestion to keep this file lean: move all ingestion framework specific things to another file. Or simply only keep traits here and move all implementations to their corresponding files

Reply:
will do the move in another PR to reduce conflict with Bridger's PR